crates/sui-metrics/src/lib.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::{hash_map::Entry, HashMap};
+use std::sync::Mutex;
+
+use once_cell::sync::OnceCell;
+use prometheus::{register_int_gauge_with_registry, IntGauge, Registry};
+use tracing::error;
+use tracing::warn;
+
+pub use scopeguard;
+
+static METRICS: OnceCell<(Registry, Mutex<HashMap<String, IntGauge>>)> = OnceCell::new();
+
+pub fn init_metrics(registry: Registry) {
+    METRICS
+        .set((registry, Default::default()))
+        .expect("sui_metrics::init_metrics duplicate init")
+}
+
+pub fn get_metric(name: String, description: String) -> Option<IntGauge> {
+    let (registry, task_metrics) = match METRICS.get() {
+        Some(v) => v,
+        None => {
+            if !cfg!(test) {
+                error!("sui-metrics not initialized");
+            }
+            return None;
+        }
+    };
+
+    let mut map = task_metrics.lock().unwrap();
+
+    match map.entry(name) {
+        Entry::Occupied(e) => Some(e.get().clone()),
+        Entry::Vacant(e) => {
+            error!("registering: {}", e.key());
+            match register_int_gauge_with_registry!(e.key(), &description, registry) {
+                Ok(metric) => {
+                    e.insert(metric.clone());
+                    Some(metric)
+                }
+                Err(err) => {
+                    warn!(name = ?e.key(), ?description, "Failed to register metric: {}", err);
+                    None
+                }
+            }
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! monitored_future {
+    ($fut: expr) => {{
+        monitored_future!("<no description>", $fut)
+    }};
+
+    ($description: expr, $fut: expr) => {{
+        let file = file!()
+            .replace("/", "_")
+            .replace("-", "_")
+            .replace(".", "_");
+
+        let name = format!("monitored_future_{}_{}", file, ::std::line!());
```

Comment:
I think it would be more reasonable to use prometheus labels, rather then compose a name.

E.g. have a `monitored_future` vec_gauge, and have labels for file and line.
This would simplify filtering in the prometheus (and is equivalent as far as performance concerned)

Reply:
Great, thanks. I don't understand prometheus very well but this sounds like the right thing to do.

Reply:
fixed, this is much much better now.